### PR TITLE
Always cache renditions

### PR DESCRIPTION
**Pull request by RealOrangeOne**
_Please check the following:_

-   [ ] Do the tests still pass?[^1]
-   [ ] Does the code comply with the style guide?
    -   [ ] Run `make lint` from the Wagtail root.
-   [ ] For Python changes: Have you added tests to cover the new/fixed behaviour?
-   [ ] For front-end changes: Did you test on all of Wagtail’s supported environments?[^2]
    -   [ ] **Please list the exact browser and operating system versions you tested**:
    -   [ ] **Please list which assistive technologies [^3] you tested**:
-   [ ] For new features: Has the documentation been updated accordingly?

Caches are incredibly useful for improving the performance of a site. We already cache lots of things under the hood of Wagtail. 

This should provide a big performance boost to those who don't set the cache, with minimal extra overhead.

[^1]: [Development Testing](https://docs.wagtail.org/en/latest/contributing/developing.html#testing)
[^2]: [Browser and device support](https://docs.wagtail.org/en/latest/contributing/developing.html#browser-and-device-support)
[^3]: [Accessibility Target](https://docs.wagtail.org/en/latest/contributing/developing.html#accessibility-targets)

### CODE REVIEW

**Comment by zerolab** [comment]
The test failures indicate these changes may not have the effect you'd expect them to

**Comment by RealOrangeOne** [comment]
It's as expected. It's just that because the DB cache is used in tests, it's showing as extra queries. Will move it to something more in-memory.

File: wagtail/images/models.py [hunk 1]
```diff
@@ -918,15 +916,15 @@ def check(cls, **kwargs):
         return errors
 
     @staticmethod
-    def construct_cache_key(image_id, filter_cache_key, filter_spec):
-        return "image-{}-{}-{}".format(image_id, filter_cache_key, filter_spec)
+    def construct_cache_key(image, filter_cache_key, filter_spec):
+        return "image-{}-{}-{}".format(image.file_hash, filter_cache_key, filter_spec)
```
Comment (RealOrangeOne) [inline wagtail/images/models.py:920]: Note to reader: This change will invalidate existing rendition caches. Do we want to put a note about this somewhere?

Comment (zerolab) [reply wagtail/images/models.py]: I think that should be fine. We can mention it in the release notes line

File: wagtail/images/models.py [hunk 2]
```diff
@@ -918,19 +910,25 @@ def check(cls, **kwargs):
...
-    def construct_cache_key(image_id, filter_cache_key, filter_spec):
-        return "image-{}-{}-{}".format(image_id, filter_cache_key, filter_spec)
+    def construct_cache_key(image, filter_cache_key, filter_spec):
+        return "wagtail-image-" + "-".join(
+            [str(image.id), image.file_hash, filter_cache_key, filter_spec]
+        )
 
-    def purge_from_cache(self):
+    @staticmethod
+    def _get_cache():
```
Comment (zerolab) [inline wagtail/images/models.py:919]: #9550 has a `renditions_cache` property which does mostly the same as this

File: wagtail/images/models.py [hunk 3]
```diff
@@ -478,15 +474,11 @@ def find_existing_rendition(self, filter: "Filter") -> "AbstractRendition":
...
-            cached_rendition = cache.get(key)
-            if cached_rendition:
-                return cached_rendition
-        except InvalidCacheBackendError:
-            pass
+        # Next, query the cache
+        cache_key = Rendition.construct_cache_key(self, cache_key, filter.spec)
+        cached_rendition = Rendition._get_cache().get(cache_key)
+        if cached_rendition:
+            return cached_rendition
```
Comment (zerolab) [inline wagtail/images/models.py:481]: for example this becomes https://github.com/wagtail/wagtail/pull/9550/files#diff-c195bc8bd4d5a11ae02d0451edb4dd4a8e7b0014398a384a1be3ff3c355a5ee8R477 in #9550

**Review by zerolab — COMMENTED** [review]
@RealOrangeOne this is most useful, but it overlaps heavily with #9550 which it makes sense to get in first (I am currently reviewing it)

File: docs/advanced_topics/performance.md [hunk 4]
```diff
@@ -24,25 +24,7 @@ CACHES = {
...
-        'BACKEND': 'django.core.cache.backends.memcached.MemcachedCache',
-        'LOCATION': '127.0.0.1:11211',
-        'TIMEOUT': 600,
-        'OPTIONS': {
-            'MAX_ENTRIES': 1000
-        }
-    }
-}
-```
+To use a different cache backend for [caching image renditions](caching_image_renditions), configure the "renditions" backend.
```
Comment (zerolab) [inline docs/advanced_topics/performance.md:27]: suggestion: leave the code snippet as it is useful (I know we could cross-link to the Django docs, but this is in-situ and relevant)

**Review by zerolab — APPROVED** [review]
This LGTM, thank you @RealOrangeOne.

I've left one suggestion (about the different cache backend code snippet), but that is non-blocking

I'm waiting for a confirmation from another core member before I merge.

---
Aside: `LocMemCache` and `./manage.py wagtail_update_image_renditions --purge-only` do not play nicely together, at least on my system, resulting in some odd behaviour whereas it would try to serve renditions from cache but they do not exist. I forgot I was using `LocMemCache` and went down a wild rabbit hole 😂

File: wagtail/images/models.py [hunk 5]
```diff
@@ -1126,19 +1114,25 @@ def check(cls, **kwargs):
...
-    def construct_cache_key(image_id, filter_cache_key, filter_spec):
-        return "image-{}-{}-{}".format(image_id, filter_cache_key, filter_spec)
+    def construct_cache_key(image, filter_cache_key, filter_spec):
+        return "wagtail-rendition-" + "-".join(
+            [str(image.id), image.file_hash, filter_cache_key, filter_spec]
+        )
 
-    def purge_from_cache(self):
+    @classproperty
+    def cache_backend(cls) -> BaseCache:
```
[the remark this answers is not in the document]
Comment (RealOrangeOne) [reply wagtail/images/models.py:1123]: I originally had `cache`, but renamed it so it's clearer https://github.com/wagtail/wagtail/pull/10612/commits/a70a782111d0858bb03813fa06e1d136fc685579

**Review by zerolab — APPROVED** [review]

**Comment by zerolab** [comment]
Merged in 6345c8c + parents

**Pull request closed without merging**